### PR TITLE
Upstreaming all features from pierrenodet/spark-smile

### PR DESCRIPTION
**Pull request by pierrenodet**
# This pull request contains three features and one enhancement

## Implicits for conversions between SMILE and Spark Dataset

ATM in this repository one need to use ```SparkDataFrame.apply``` from ```smile.data``` to actually convert a Spark DataFrame to a SMILE DataFrame, now by importing ```scala.data._``` (or ```scala.data.SparkDataFrameOps```) a new extension method ```toSmileDF``` is available on Spark DataFrame.

**Example :**
```scala
import scala.data._
val sparkMushrooms = spark.read.format("libsvm").load("data/mushrooms.svm")
val smileMushrooms = sparkMushrooms.toSmileDF()
```

I don't know how you organize code in this repository, but I used to store this extension method in ```org.apache.spark.smile.implicits._```, it depends on what you feel should be the best for the user.

## Conversion from SMILE dataset to Spark Dataset

Same feature as above but the other way around. To way to use it either with ```SmileDataFrame.apply(sparkDF)``` from ```smile.data``` or ```scala.data.SmileDataFrameOps```

**Example :**
```scala
import scala.data._
val spark = SparkSession.builder().master("local[*]").getOrCreate()

val smileMushrooms = read.arff("data/mushrooms.arff").omitNullRows().toSparkDF(spark)
val sparkMushrooms = smileMushrooms.toSparkDF(spark)
val otherSparkMushrooms =SmileDataFrame(smileMushrooms,spark)

mushrooms.show()
```

I feel like the naming of SparkDataFrame and SmileDataFrame is unfortunate, they might be confusing for the user.

## Helpers to use SMILE Regressor and SMILE Classifier in Spark MLLib Pipelines

This features create two classes ```SmileClassifer``` and ```SmileRegressor``` that wraps a smile classification and regression model in a Spark Estimator.

**Example :**
```scala
import org.apache.spark.ml.classification.SmileClassifier

val raw = spark.read.format("libsvm").load("data/mushrooms.svm")

val scl = new SmileClassifier()
  .setTrainer({ (x, y) => knn(x, y, 3) })

val bce = new BinaryClassificationEvaluator()
  .setLabelCol("label")
  .setRawPredictionCol("rawPrediction")

val model = scl.fit(data)

println(bce.evaluate(model.transform(data)))

model.write.overwrite().save("/tmp/model")
val loaded = SmileClassificationModel.load("/tmp/model")
println(bce.evaluate(loaded.transform(data)))
```

## Helpers for Distributed GridSearchCrossValidation over Spark Executors

This features does cross validation of multiple trainers over all spark executors from the spark job. This helper function resides in ```scala.tuning.sparkgscv```.

**Example :**
```scala
import scala.tuning._

val spark = SparkSession.builder().master("local[*]").getOrCreate()

val mushrooms = read.arff("data/mushrooms.arff")

val x = mushrooms.select(1,22).toArray
val y = mushrooms("class").toIntArray

val knn3 = (x:Array[Array[Double]], y:Array[Int]) => KNN.fit(x, y, 3)
val knn5 = (x:Array[Array[Double]], y:Array[Int]) => KNN.fit(x, y, 5)

val res = sparkgscv(spark)(5, x, y, Seq(new Accuracy()): _*) (Seq(knn3,knn5):_*)

println(res(0))
```

Tell me if you prefer spliting this in multiple PR.

### CODE REVIEW

File: spark/src/main/scala/org/apache/spark/ml/classification/SmileClassifier.scala [hunk 1]
```diff
@@ -0,0 +1,238 @@
+package org.apache.spark.ml.classification
```
Comment (haifengl) [inline spark/src/main/scala/org/apache/spark/ml/classification/SmileClassifier.scala:1]: I wouldn't pollute spark's namespace.

File: spark/build.sbt [hunk 2]
```diff
@@ -1,14 +1,21 @@
...
+fork.in(Test) := true
+parallelExecution.in(Test) := false
+
 libraryDependencies ++= {
   val sparkV = "2.4.7"
   Seq(
     "org.apache.spark" %% "spark-core" % sparkV % Provided,
-    "org.apache.spark" %% "spark-sql"  % sparkV % Provided
+    "org.apache.spark" %% "spark-sql"  % sparkV % Provided,
+    "org.apache.spark" %% "spark-mllib"  % sparkV % Provided,
```
Comment (haifengl) [inline spark/build.sbt:18]: What SparkSQL features are needed?

File: spark/src/main/scala/smile/data/SparkDataFrame.scala [hunk 3]
```diff
@@ -1,35 +1,23 @@
-/*******************************************************************************
```
Comment (haifengl) [inline spark/src/main/scala/smile/data/SparkDataFrame.scala:1]: Please keep the copyright header (for all files).

File: .gitignore [hunk 4]
```diff
@@ -40,3 +40,9 @@ hs_err_pid*
 .cache-tests
 .classpath
 */.gitignore
+
```
[the remark this answers is not in the document]
Comment (pierrenodet) [reply .gitignore:43]: yep my bad, doing it

File: spark/build.sbt [hunk 5]
```diff
@@ -1,14 +1,21 @@
 name := "smile-spark"
 
+scalaVersion := "2.12.8"
```
[the remark this answers is not in the document]
Comment (pierrenodet) [reply spark/build.sbt:3]: ok should on which branch should i push ? there is no spark versions yet for scala 2.13

File: spark/src/main/scala/smile/data/SparkDataFrame.scala (hunk 3 above)
Comment (pierrenodet) [reply spark/src/main/scala/smile/data/SparkDataFrame.scala:1]: yep my bad, btw https://github.com/sbt/sbt-header could be used to make this process automatic

File: spark/src/main/scala/org/apache/spark/ml/classification/SmileClassifier.scala (hunk 1 above)
Comment (pierrenodet) [reply spark/src/main/scala/org/apache/spark/ml/classification/SmileClassifier.scala:1]: yes but there is a lot of useful code which is private spark I use to make the code more robust/production grade, do you prefer that i copy paste what I need from spark ? I have used the same approach as : https://github.com/Azure/mmlspark for namespaces

File: spark/build.sbt (hunk 2 above)
Comment (pierrenodet) [reply spark/build.sbt:18]: The project was already depending on spark-sql and we need it for Row, DataFrame, etc. And Spark MLLib depends on SparkSQL anyway

**Comment by pierrenodet** [comment]
@haifengl I have taken into account your comments

File: spark/src/main/scala/org/apache/spark/ml/regression/SmileRegressor.scala [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+import org.apache.spark.ml.util.Instrumentation._
+import org.apache.spark.ml.util.{Identifiable, _}
+import org.apache.spark.ml.{Predictor, PredictorParams}
+import org.apache.spark.sql.Dataset
+import org.apache.spark.storage.StorageLevel
+import org.json4s.JObject
+import org.json4s.JsonDSL._
+import org.json4s.jackson.JsonMethods._
+
+private[ml] trait SmileRegressorParams extends Params with PredictorParams {
```
Comment (haifengl) [inline spark/src/main/scala/org/apache/spark/ml/regression/SmileRegressor.scala:35]: I don't see difference between SmileClassificationParams and SmileRegressorParams. Do we need both?

File: spark/src/main/scala/org/apache/spark/ml/regression/SmileRegressor.scala [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+                path: String,
+                sc: SparkContext,
+                expectedClassName: String): DefaultParamsReader.Metadata = {
+
+    DefaultParamsReader.loadMetadata(path, sc, expectedClassName)
+  }
+
+}
+
+class SmileRegressor(override val uid: String)
```
Comment (haifengl) [inline spark/src/main/scala/org/apache/spark/ml/regression/SmileRegressor.scala:81]: Regressor usually refers to the independent variables (X), not the model. It is better named as SmileRegression.

File: spark/src/main/scala/org/apache/spark/smile/SparkDataTypes.scala [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+package org.apache.spark.smile
+
+import smile.data.`type`.{DataType, DataTypes, StructField, StructType}
+import org.apache.spark.sql.types._
+import org.apache.spark.ml.linalg.VectorUDT
+import org.apache.spark.mllib.linalg.{VectorUDT => OldVectorUDT}
+import org.apache.spark.sql.catalyst.encoders.ExpressionEncoder
+
+object SparkDataTypes {
```
Comment (haifengl) [inline spark/src/main/scala/org/apache/spark/smile/SparkDataTypes.scala:26]: I guess that classes in org.apache.spark.smile package don't use spark's private features. If so, let's move them to smile namespace.

File: spark/src/main/scala/smile/data/SparkDataFrame.scala [hunk 9]
```diff
@@ -45,53 +50,9 @@ case class SparkRowTuple(row: org.apache.spark.sql.Row, val schema: StructType)
...
-    val schema = SparkDataTypes.smileSchema(df.schema)
-    DataFrame.of(
-      df.collect()
-        .map(row => SparkRowTuple(row, schema))
-        .toList
-        .asJava)
-  }
-}
-
-object SparkDataTypes {
```
Comment (haifengl) [inline spark/src/main/scala/smile/data/SparkDataFrame.scala:66]: Why are these removed?

File: spark/src/main/scala/smile/data/package.scala [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Smile.  If not, see <https://www.gnu.org/licenses/>.
+ ******************************************************************************/
+
+package smile
+
+import org.apache.spark.sql.SparkSession
+
+package object data {
```
Comment (haifengl) [inline spark/src/main/scala/smile/data/package.scala:22]: It is better to keep these in spark package object. The `data` package should not depend on spark.

File: spark/src/main/scala/smile/tuning/package.scala [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+import java.util.function.BiFunction
+
+import org.apache.spark.sql.SparkSession
+import smile.classification.Classifier
+import smile.validation.{Accuracy, ClassificationMeasure, CrossValidation}
+
+import scala.reflect.ClassTag
+
+case class SerializableClassificationMeasure(@transient measure: ClassificationMeasure)
```
Comment (haifengl) [inline spark/src/main/scala/smile/tuning/package.scala:28]: Please move this to smile.spark package too. It is not good to spread spark code to many packages. Not everyone uses spark.

File: spark/src/main/scala/smile/data/SmileDataFrame.scala [hunk 12]
```diff
@@ -0,0 +1,40 @@
+/*******************************************************************************
```
Comment (haifengl) [inline spark/src/main/scala/smile/data/SmileDataFrame.scala:1]: I will move this file to smile.spark package as we have more spark related facilities now. It is better to keep them in a separated namespace.

**Comment by haifengl** [comment]
@pierrenodet Thanks! I add more comments.

File: spark/src/test/resources/mushrooms.arff [hunk 13]
```diff
@@ -0,0 +1,8301 @@
+% 1. Title: Mushroom Database
```
Comment (haifengl) [inline spark/src/test/resources/mushrooms.arff:1]: All the test data are in `shell/src/universal/data` directory.

File: spark/src/test/resources/mushrooms.svm [hunk 14]
```diff
@@ -0,0 +1,8124 @@
+1 6:1 8:1 15:1 21:1 29:1 33:1 34:1 37:1 42:1 50:1 53:1 57:1 67:1 76:1 78:1 81:1 84:1 86:1 93:1 103:1 111:1
```
Comment (haifengl) [inline spark/src/test/resources/mushrooms.svm]: `shell/src/universal/data/libsvm` directory.

Comment (pierrenodet) [reply spark/src/test/resources/mushrooms.svm]: done

File: spark/src/test/resources/mushrooms.arff (hunk 13 above)
Comment (pierrenodet) [reply spark/src/test/resources/mushrooms.arff]: done

File: spark/src/main/scala/smile/data/SmileDataFrame.scala (hunk 12 above)
Comment (pierrenodet) [reply spark/src/main/scala/smile/data/SmileDataFrame.scala]: ok

File: spark/src/main/scala/smile/tuning/package.scala (hunk 11 above)
Comment (pierrenodet) [reply spark/src/main/scala/smile/tuning/package.scala]: done in smile.spark.tuning, do you prefer smile.spark.tuning or just smile.spark ?

File: spark/src/main/scala/smile/data/package.scala (hunk 10 above)
Comment (pierrenodet) [reply spark/src/main/scala/smile/data/package.scala]: done

File: spark/src/main/scala/smile/data/SparkDataFrame.scala (hunk 9 above)
Comment (pierrenodet) [reply spark/src/main/scala/smile/data/SparkDataFrame.scala:66]: I removed the implicits conversions because I thought it was not a good scala practice and it's probably never going to be used by users directly. Theses methods are still present here : spark/src/main/scala/org/apache/spark/smile/SparkDataTypes.scala as actual methods

File: spark/src/main/scala/org/apache/spark/smile/SparkDataTypes.scala (hunk 8 above)
Comment (pierrenodet) [reply spark/src/main/scala/org/apache/spark/smile/SparkDataTypes.scala:30]: VectorUDT (which is the features DataType used by Spark Estimators) is private Spark as the UserDefineType.

File: spark/src/main/scala/org/apache/spark/ml/regression/SmileRegressor.scala (hunk 7 above)
Comment (pierrenodet) [reply spark/src/main/scala/org/apache/spark/ml/regression/SmileRegressor.scala]: That's the usual naming convention in Spark MLLib with RandomForestRegressor and FMRegressor : https://spark.apache.org/docs/latest/ml-classification-regression.html#factorization-machines-regressor but as SurvivalRegression or IsotonicRegression is used to i'm going to change it as SmileRegression.

File: spark/src/main/scala/org/apache/spark/ml/regression/SmileRegressor.scala (hunk 6 above)
Comment (pierrenodet) [reply spark/src/main/scala/org/apache/spark/ml/regression/SmileRegressor.scala]: It's different because the trainer params is either ... => smile.regression.Regression[Array[Double]] or ... =>  smile.classification.Classifier[Array[Double]]

File: spark/src/main/scala/org/apache/spark/ml/classification/SmileClassifier.scala [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+import org.apache.spark.ml.param._
+import org.apache.spark.ml.util.Instrumentation._
+import org.apache.spark.ml.util.{Identifiable, _}
+import org.apache.spark.sql.Dataset
+import org.apache.spark.storage.StorageLevel
+import org.json4s.JsonDSL._
+import org.json4s.jackson.JsonMethods._
+import org.json4s.{DefaultFormats, JObject}
+
+private[ml] trait SmileClassifierParams extends Params with ClassifierParams {
```
[the remark this answers is not in the document]
Comment (pierrenodet) [reply spark/src/main/scala/org/apache/spark/ml/classification/SmileClassifier.scala:39]: done, tell me if it's not enough

File: spark/build.sbt [hunk 16]
```diff
@@ -5,10 +5,15 @@ name := "smile-spark"
 crossPaths := true
 autoScalaLibrary := true
 
+fork.in(Test) := true
```
[the remark this answers is not in the document]
Comment (pierrenodet) [reply spark/build.sbt]: ok removed

File: spark/build.sbt [hunk 17]
```diff
@@ -5,10 +5,15 @@ name := "smile-spark"
...
+parallelExecution.in(Test) := false
```
[the remark this answers is not in the document]
Comment (pierrenodet) [reply spark/build.sbt]: ok removed

**Comment by pierrenodet** [comment]
@haifengl I have taken into account your recent comments. Thank you to take time to make these reviews !

**Comment by haifengl** [comment]
There is a conflict as I already move SparkDataFrame to smile.spark package. Can you please fix it so that I can merge? Thanks.

**Comment by pierrenodet** [comment]
@haifengl I merged master branch into this one so the conflicts should be resolved

**Comment by haifengl** [comment]
Thank you very much!
